### PR TITLE
support rebar3 resource type git_subdir

### PR DESCRIPTION
**Pull request by tsloughter**
Related rebar3 PR which add git_subdir: https://github.com/erlang/rebar3/pull/2239

Hoping to get feedback while I figure out adding tests.

### CODE REVIEW

File: lib/mix/lib/mix/rebar.ex [hunk 1]
```diff
@@ -151,6 +151,21 @@ defmodule Mix.Rebar do
...
+            [_, sparse_dir | _] ->
+              {:git, [{:sparse, sparse_dir}]}
+
+            _ ->
+              {:git, []}
+          end
+
+        _ ->
+          {:git, []}
+      end
```
Comment (josevalim) [inline lib/mix/lib/mix/rebar.ex:167]:
```suggestion
    {scm, additional_opts} =
      case {scm, source} do
        {:git_subdir, [_, sparse_dir | _]} -> {:git, [{:sparse, sparse_dir}]}
        {_, _} -> {:git, []}
      end
```

Comment (josevalim) [reply lib/mix/lib/mix/rebar.ex:167]: Also, couldn't someone try to use git_subdir with branch/tag/ref? In this case, shouldn't this case also return the new source? Something like this:

```suggestion
    {scm, additional_opts, source} =
      case {scm, source} do
        {:git_subdir, [_, sparse_dir | source]} -> {:git, [sparse: sparse_dir], source}
        {_, _} -> {:git, [], source}
      end
```

?

Comment (tsloughter) [reply lib/mix/lib/mix/rebar.ex:167]: It has to include a branch/tag/ref, that is why the head of the source is skipped and then parsed in the next `case`.

Comment (tsloughter) [reply lib/mix/lib/mix/rebar.ex:167]: Format is: `{git_subdir, Url, BranchTagRef, Dir}`.

Comment (josevalim) [reply lib/mix/lib/mix/rebar.ex]: ahhhhhhhhhhh, I see! So ignore my second suggestion, although I think my first suggestion is still good, right?

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**